cnd/src/swap_protocols/metadata_store.rs

```diff
@@ -85,49 +82,86 @@ impl Metadata {
     }
 }
 
-#[derive(Debug, Fail)]
+#[derive(Debug)]
 pub enum Error {
```

Comment:
Usually, we only store the source of the error inside the variant. This would potentially leak some information here so I am not quite sure. Would have to investigate in detail.

If you can't pass on the source error, I'd still be in favor of just putting relevant data into the variant and not an already formatted string. Try to delay the formatting for as long as possible (you don't know how your caller wants to handle the error, let alone localization etc).

Reply:
The error handling in general needs work, it was best effort at the time - I'd like to spend some time on it.  There are a bunch of things that I'm not happy with in this PR, finding the balance of getting something merged and working towards perfection is not easy here especially considering the complexity of this task compared to my Rust-foo.  I'm not going to merge this right away, open to suggestions on things that you think should be done before merge and things that you think are ok to do after merge.  I'll probably go for 'loose consensus and running code' as they say.

Reply:
Yes, should have probably made this more explicit: this comment was definitely not a blocker :)